packages/ledger/src/index.ts

```diff
@@ -0,0 +1,339 @@
+  )
+  return {
+    derivationPath: `${derivationPath}/${index}`,
+    address,
+    balance: {
+      asset: asset.label,
+      value: await getBalance(address, rpcUrl)
+    }
+  }
+}
+
+const getAddresses = (
+  account: LedgerAccount,
+  asset: Asset,
+  currentChain: Chain,
+  offset: number = 0,
+  limit: number = 5
+): Promise<Account[]> =>
+  limit - offset <= 0
+    ? (() => {
+        throw new Error('Offset must be less than limit')
+      })()
+    : Promise.all(
+        Array.from({ length: limit - offset }, (_, index) =>
+          getAddress(account, asset, currentChain, index)
+        )
+      )
+
+function ledger({
+  customNetwork
+}: {
+  customNetwork?: CustomNetwork
+} = {}): WalletInit {
+  const getIcon = async () => (await import('./icon')).default
+  return () => {
+    let accounts: Account[] | undefined
+    return {
+      label: 'Ledger',
+      getIcon,
+      getInterface: async ({ EventEmitter, chains }) => {
+        const Eth = (await import('@ledgerhq/hw-app-eth')).default
+        const { TransactionFactory: Transaction, Capability } = await import(
+          '@ethereumjs/tx'
+        )
+        const { default: Common, Hardfork } = await import('@ethereumjs/common')
+        const { compress } = (await import('eth-crypto')).publicKey
+        const ethUtil = await import('ethereumjs-util')
+        const { getStructHash, getMessage } = await import('eip-712')
+
+        const transport: Transport = await getTransport()
+        const eth = new Eth(transport)
+        const eventEmitter = new EventEmitter()
+
+        let currentChain: Chain = chains[0]
+        const scanAccounts = async ({
+          derivationPath,
+          chainId,
+          asset
+        }: ScanAccountsOptions): Promise<Account[]> => {
+          currentChain = chains.find(({ id }) => id === chainId) ?? currentChain
+
+          const { publicKey, chainCode } = await eth.getAddress(
+            derivationPath,
+            false,
+            true // set to true to return chainCode
+          )
+
+          return getAddresses(
+            {
+              publicKey: compress(publicKey),
+              chainCode: chainCode ?? '',
+              derivationPath
+            },
+            asset,
+            currentChain
+          )
+        }
+
+        const getAccounts = async () => {
+          return (
+            accounts ??
+            (accounts = await accountSelect({
+              basePaths: DEFAULT_BASE_PATHS,
+              assets,
+              chains,
+              scanAccounts,
+              walletIcon: await getIcon()
+            }))
+          )
+        }
+
+        const ledgerProvider = {}
+
+        const provider = createEIP1193Provider(ledgerProvider, {
+          eth_requestAccounts: async baseRequest => {
+            // Triggers the account select modal if no accounts have been selected
+            const [{ address }] = await getAccounts()
```

Comment:
If the user exits the account select modal without selecting an account, then the `getAccounts` function will return an empty array. I think we should add a check to make sure the result has `length`, and if not then we need to throw the error code for when the user rejects connection to be compliant with the `eth_requestAccounts` spec. In the error case, we will also need to see the `account` variable to undefined to indicate that accounts have not yet been connected.